### PR TITLE
feat: facecard in gamestate api

### LINKED ISSUE #948: [Feature]: Implement Playing FaceCards (Kings, Queens, and glasses 8s) in the GameState API
### Feature Summary

We need to implement playing Face Cards (Kings, Queens, Glasses Eights; Jacks to be handled separately) in the GameState API in order to support the full game logic in this API. See the [spec](https://docs.google.com/document/d/1vYe4xoVULIbbFU2yhErok-Qm1cg8ujMGNqlUOgNXLwE/edit#heading=h.wpeuh0qc6ps2) for a general overview of the API, its underlying motivations, and the development approach.

Playing a FaceCard means moving the `playedCard` from the requesting player's `hand` to their `faceCards`. This should also increment the turn, and set the `phase` to `GamePhase.MAIN`

### Detailed Description

Implementation should take the general approach shown in the [Backend Control Flow section of the GameState API Spec](https://docs.google.com/document/d/1vYe4xoVULIbbFU2yhErok-Qm1cg8ujMGNqlUOgNXLwE/edit#heading=h.vgf103gono5r) and follow the example of #1025 

Making this move, like all others, will be triggered via the endpoint:

```
POST /api/game/:gameId/move
```
Which move gets made is determined by the request body, which should specify a moveType field with a value from the MoveType enum. Drawing a card will be triggered by sending a request body like:

```
{ moveType: MoveType.FACECARD, playedCard: 'KH'};
```
When a `moveType: MoveType.FACECARD` is sent in the req body, the endpoint will attempt to utilize the `faceCard` helpers: `sails.helpers.gamestate.moves.faceCard.validate()` and `sails.helpers.gamestate.moves.faceCard.execute()`. Each of these should follow the paradigm outlined in the [docs](https://docs.google.com/document/d/1vYe4xoVULIbbFU2yhErok-Qm1cg8ujMGNqlUOgNXLwE/edit#heading=h.vgf103gono5r):

There are four pieces to implement:
1. update the `api/policies/hasValidMoveBody` policy to allow and validate requests with a moveType of `MoveType.FACECARD`
2. Implement the `api/helpers/gamestate/moves/faceCard/validate` helper to verify if the move is legal
3. Implement the `api/helpers/gamestate/moves/faceCard/execute` helper to create a new GameState that plays a facecard for the requested player
4. Unskip the relevant tests by removing the `cy.skipOnOnGameStateApi()` commands from tests which can now pass in the GameState API

### Updating the `api/policies/hasValidMoveBody` policy
We must also update the policy `api/policies/hasValidMoveBody.js` to accommodate the `MoveType.FACECARD` moveType. This policy is already applied to the endpoint for making a move and it uses a `switch()` statement on `req.body.moveType` to validate that a given req body has all the data necessary for the given `moveType`. We should add a `MoveType.FACECARD` case to this policy.

A `moveType`  of `MoveType.FACECARD` requires a `playedCardId` which is a valid card id. If `playedCardId` is present in req.body and it can be found in `utils/DeckIds.json`, then we can `return res.next()` in this switch case. Otherwise we should return `res.badRequest()` with an error that explains which `req.body` attribute is missing or invalid.

### Implementing `validate()`
Create a new `api/helpers/gamestate/moves/faceCard/validate.js` helper file that matches the general setup of the existing `api/helpers/gamestate/moves/points/validate.js` helper. It should throw errors if the requested move is illegal. It should have three inputs: 
1. `currentState` - the unpacked `GameState` from before the move is made
2. `requestedMove` -req body describing the move being made3.
3. `playedBy` - 0 or 1 for whether p0 or p1 is making the move.
`validate()` should throw if:
* It is not the requesting player's turn
* The GameState's `phase !== GamePhase.MAIN`
* The `playedCard` is not in the player's hand (`throw new Error("game.snackbar.global.playFromHand")`)
* The `playedCard` is frozen (`throw new Error('game.snackbar.global.cardFrozen')`)
* The `playedCard` has a rank other than 8, 12, or 13 (`throw new Error('game.snackbar.faceCard.withoutTarget'`)

### Implementing `execute()`
Create a new `api/helpers/gamestate/moves/faceCard/execute.js` helper file that matches the general approach of the existing execute helper `api/helpers/gamestate/moves/faceCard/execute.js`. It should create and return a new `GameState` that is the result of making the `faceCard` move for the requested player i.e. **it should remove `playedCard` from the player's hand, and add it to their `faceCards`. It should also increment the turn and set the phase to `GamePhase.MAIN`. The `playedBy`, and `playedCard` attributes should be included in the returned `GameState`.

`execute()`  should have 3 inputs:
1. `currentState` - the unpacked `GameState` from before the move is made
2. `requestedMove` -req body describing the move being made
3. `playedBy` - 0 or 1 for whether p0 or p1 is making the move.

### Unskipping Tests
Because we are approaching the `GameState API project with a feature-flag approach, all changes are merged into the `main` branch and the application and tests are set to dynamically run in either the GameState API mode, or the legacy mode, based on the `VITE_USE_GAMESTATE_API` env variable. In CI, we run the full suite of tests against the legacy version of the application and then run a subset of the tests against the GameState API version of the application.

Which tests are run in the GameState API version of CI is controlled via the test command `cy.skipOnGameStateApi()` which programmatically skips a test if the env var `VITE_USE_GAMESTATE_API` is enabled. As we develop new features in the GameState API, we should remove the `cy.skipOnGameStateApi()` command from more and more of the tests.

**Essential tests**
For this PR, at minimum, we should unskip the following tests, which should be able to pass once it is working correctly. All the following tests are in the `tests/e2e/specs/in-game` folder:
* `basicMoves.spec.js` - `it('Plays kings')`
* `basicMoves.spec.js` - `it('Plays queens')`
* `basicMoves.spec.js` - `it('Plays eights for glasses')`

You can run the tests against the GameState API locally by:
 booting the server in gamestate mode with
`npm run start:server:gamestate`

Then booting the client in gamestate mode with
`npm run start:client:gamestate`

running the tests in gamestate mode with
`npm run e2e:client:gamestate`.

### PR DESCRIPTION
**Pull request by haaarsh619**
## Issue number #948

Relevant [#948](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)
- Resolves #948

## Please check the following

- [X] Do the tests still pass? (see [Run the Tests](https://github.com/cuttle-cards/cuttle/blob/main/docs/setup-development.md#run-the-tests))
- [X] Is the code formatted properly? (see [Linting (Formatting)](https://github.com/cuttle-cards/cuttle/blob/main/docs/setup-development.md#linting-formatting))
- For New Features:
  - [ ] Have tests been added to cover any new features or fixes?
  - [ ] Has the documentation been updated accordingly?

### CODE REVIEW

File: api/helpers/gamestate/moves/faceCard/execute.js [hunk 1]
```diff
@@ -0,0 +1,51 @@
+const GamePhase = require('../../../../../utils/GamePhase.json');
+
+module.exports = {
+  friendlyName: 'Play Face Card',
+
+  description: 'Returns new GameState resulting from requested Face Card  move',
```
Comment (itsalaidbacklife) [inline api/helpers/gamestate/moves/faceCard/execute.js:6]:
```suggestion
  description: 'Returns new GameState resulting from requested Face Card move',
```

File: api/helpers/gamestate/moves/faceCard/execute.js [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+  inputs: {
+    currentState: {
+      type: 'ref',
+      description: 'The latest GameState before the requesting player plays a face card',
+      required: true,
+    },
+    /**
+     * @param { Object } requestedMove - Object describing the request to play points (req.body)
+     * @param { String } requestedMove.cardId - Card Played for points
```
Comment (itsalaidbacklife) [inline api/helpers/gamestate/moves/faceCard/execute.js:16]:
```suggestion
     * @param { Object } requestedMove - Object describing the request to play a face card (req.body)
     * @param { String } requestedMove.cardId - Card Played being played as face card
```

File: api/helpers/gamestate/moves/faceCard/execute.js [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+  sync: true, // synchronous helper
+  fn: ({ currentState, requestedMove, playedBy }, exits) => {
+    const { cardId } = requestedMove;
+    let result = _.cloneDeep(currentState);
+
+    const player = playedBy ? result.p1 : result.p0;
+    const cardIndex = player.hand.findIndex(({ id }) => id === cardId);
+
+    const [playedCard] = player.hand.splice(cardIndex, 1);
+    player.push(playedCard);
```
Comment (itsalaidbacklife) [inline api/helpers/gamestate/moves/faceCard/execute.js:37]: `player` (which comes from `currentState.p0` or `currentState.p1` is a Player object which looks like:

```
{
  hand: Card[];
  points: Card[];
  faceCards: Card[]
}
```

So we need to remove the card from the player's hand (which you're doing correctly) and add it to their **faceCards**, not the player themself
```suggestion
    player.faceCards.push(playedCard);
```

File: api/helpers/gamestate/moves/faceCard/execute.js [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+  },
+  sync: true, // synchronous helper
+  fn: ({ currentState, requestedMove, playedBy }, exits) => {
+    const { cardId } = requestedMove;
+    let result = _.cloneDeep(currentState);
+
+    const player = playedBy ? result.p1 : result.p0;
+    const cardIndex = player.hand.findIndex(({ id }) => id === cardId);
+
+    const [playedCard] = player.hand.splice(cardIndex, 1);
```
Comment (itsalaidbacklife) [inline api/helpers/gamestate/moves/faceCard/execute.js:36]: nitpick: prefer whitespace around destructured variable. There's probably an eslint rule for this that would be helpful to configure in a separate issue

File: api/helpers/gamestate/moves/faceCard/validate.js [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+
+  inputs: {
+    currentState: {
+      type: 'ref',
+      descriptions: 'Object containing the current game state',
+      required: true,
+    },
+    /**
+     * @param { Object } requestedMove - Object describing the request to play points
+     * @param { String } requestedMove.cardId - Card Played for points
```
Comment (itsalaidbacklife) [inline api/helpers/gamestate/moves/faceCard/validate.js:16]:
```suggestion
     * @param { Object } requestedMove - Object describing the request to play face card
     * @param { String } requestedMove.cardId - Card Played as face card
```

File: api/helpers/gamestate/moves/faceCard/validate.js [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+  },
+  sync: true,
+  fn: ({ requestedMove, currentState, playedBy }, exits) => {
+    try {
+      const playerKey = playedBy ? 'p1' : 'p0';
+
+      const cardPlayed = currentState[playerKey].hand.find(({ id }) => id === requestedMove.cardId);
+
+      if (currentState.phase !== GamePhase.MAIN) {
+        throw new Error(`Can only play face card  in main phase, not ${currentState.phase}`);
```
Comment (itsalaidbacklife) [inline api/helpers/gamestate/moves/faceCard/validate.js:38]: whitespace
```suggestion
        throw new Error(`Can only play face card in main phase, not ${currentState.phase}`);
```

File: api/helpers/gamestate/moves/faceCard/validate.js [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+
+      if (cardPlayed.isFrozen) {
+        throw new Error('game.snackbar.global.cardFrozen');
+      }
+
+      if (currentState.turn % 2 !== playedBy) {
+        throw new Error('game.snackbar.global.notYourTurn');
+      }
+
+      if ((cardPlayed.rank >= 12 && cardPlayed.rank <= 13) || cardPlayed.rank === 8) {
```
Comment (itsalaidbacklife) [inline api/helpers/gamestate/moves/faceCard/validate.js:53]: shorter with .includes()
```suggestion
      if (![8, 12, 13].includes(cardPlayed.rank)) {
```

File: src/routes/rules/components/MultiplayerRuleBlock.vue
Comment (itsalaidbacklife) [inline src/routes/rules/components/MultiplayerRuleBlock.vue:1]: Thanks this lint warning was annoying

**Review by itsalaidbacklife — COMMENTED** [review]
Great work! This is almost ready. Mostly nitpicks and one type error with the .push() call.

**Comment by itsalaidbacklife** [comment]
@haaarsh619 can you please also unskip `it('Shows when player wins game with 0 points and four kings')` from `tests/e2e/specs/in-game/winConditions.spec.js`?

**Review by Haviles04 — APPROVED** [review]
LGTM, Nice work!

File: tests/e2e/specs/in-game/basicMoves.spec.js [hunk 8]
```diff
@@ -230,7 +229,6 @@ describe('Game Basic Moves - P0 Perspective', () => {
   });
 
   it('Plays Queens', () => {
-    cy.skipOnGameStateApi();
```
Comment (itsalaidbacklife) [inline tests/e2e/specs/in-game/basicMoves.spec.js:233]: Turns out this test can't be included yet because it requires playing a jack. Keep this as skipped

**Review by itsalaidbacklife — APPROVED** [review]
Great work! This looks ready. The failing test is for playing queens which it turns out can't be unskipped yet. Reskip that test and this is good to go

File: api/helpers/gamestate/moves/face-card/validate.js [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+      description: 'Player number of player requesting move',
+      required: true,
+    },
+  },
+  sync: true,
+  fn: ({ requestedMove, currentState, playedBy }, exits) => {
+    try {
+      const playerKey = playedBy ? 'p1' : 'p0';
+
+      const cardPlayed = currentState[playerKey].hand.find(({ id }) => id === requestedMove.cardId);
```
Comment (itsalaidbacklife) [inline api/helpers/gamestate/moves/face-card/validate.js:35]: Last request: let's switch this to using the same ternary expression as the execute() helper so it gets the player object with currentState.p0 and currentState.p1 into a variable called `player`. That will be more consistent

**Review by itsalaidbacklife — APPROVED** [review]
Great work. One final nitpick and it's ready

**Pull request merged**